_posts/2017-10-27-conventional-interfaces.md

```diff
@@ -0,0 +1,140 @@
+---
+layout: post
+name: 2017-10-26-conventional-interfaces
+title: Conventional Interfaces
+date: 2017-10-14 10:20:00 +00:00
+author: Carlos Morera
+image:
+   src: 
+abstract: TODO.
+canonical:
+    name: Carlos's blog
+    href: https://medium.com/@hkoundi/find-a-workplace-where-you-can-grow-16172a5ab957/
+tags:
+- FP
+- Haskell
+- SICP
+---
+
+# Intro
+
+* Whilst reading SICP book I discovered the Sequences as Conventional Interfaces section. I'm going to somehow remove the "Sequences" connotation as they probably use it given that the book uses lisp where everything is a sequence.
+
+# What are Conventional Interfaces
+
+SICP first define conventional interfaces as a design principle for working with data structures. 
+
+It is a set of standard operations for connecting the different steps require to implement computer programs. These operations may be familiar to you as they will probably be provided in some way in your preferred language, e.g. map, filter, flatmap (also now as bind), fold (also known as reduce), etc.. 
+
+These operations let us capture common patterns in the implementation of programs that a priori are different structurally. These common patterns let us a think about different programs in a similar way. This is incredibly important as we can easily express different programs by simply combining the same set of operations in different ways.
+
+SCIP introduces the concept of signal processing as a metaphore to reason about this programming style. Quote: "A signal-processing engineer would find it natural to conceptualize processes in terms of signals flowing through a casacade of stages, each of which implements part of the program plan". We will expand more on this metaphore later on, but for now it is important to emphasize that making the signal-flow structure evident in the design of our programs let us increase the modularity and readability of the resulting code. 
+
+It is also important to emphasize that these operations increase the level of abstraction at which we can write code by taking away low level details such as, iteration, recursion, conditional statements...
+
+Let's consider the two following examples taken from SICP and that I have translated from LISP to Haskell.
+
+* This set of operations let us extract the distints parts of a computation by decomposing the different concerns
+* The implementation of a sequence dictates how the sequence is iterated (https://softwareengineering.stackexchange.com/questions/284912/why-sequences-are-recommended-as-conventional-interfaces)
+
+# Discovering common patterns
+
+Consider the following two functions taken from the SICP book.
+
+## Sum squares of the leaves of a tree that are odd
+
+```haskell
+
+-- A simple type representing a binary tree
+data BinaryTree a =
+    Leaf a
+  | Node (BinaryTree a) (BinaryTree a) deriving (Eq, Ord, Show)
```

Comment:
Is the `deriving` part needed?

Reply:
Actually, it isn't, thanks. The reason why they were there is that I took this implementation from the Haskell book and modified it so the value is on the leaves instead of in the nodes. Their definition is as follows:

```
data BinaryTree a =
Leaf
| Node (BinaryTree a) a (BinaryTree a) deriving (Eq, Ord, Show)
```

Will modify the post to remove the unnecessary deriving :)